bliss/simulator/decoder.py

```diff
@@ -123,30 +162,53 @@ def render_galaxy(self, galaxy_params: Tensor):
             sheared_bulge = bulge.shear(q=bulge_q, beta=beta_radians * galsim.radians)
             components.append(sheared_bulge)
         galaxy = galsim.Add(components)
-        return galsim.Convolution(galaxy, self.psf_galsim)
+        return galsim.Convolution(galaxy, psf[bnd])
 
-    def render_images(self, tile_cat: TileCatalog):
+    def render_images(self, tile_cat: TileCatalog, rcf):
         batch_size, n_tiles_h, n_tiles_w = tile_cat.n_sources.shape
+        assert rcf.shape[0] == batch_size
+
         slen_h = tile_cat.tile_slen * n_tiles_h
         slen_w = tile_cat.tile_slen * n_tiles_w
         images = np.zeros((batch_size, self.n_bands, slen_h, slen_w), dtype=np.float32)
 
         full_cat = tile_cat.to_full_params()
 
+        # use the PSF from specified row/camcol/field
+        psfs = [self.psf_galsim[tuple(rcf[b])] for b in range(batch_size)]  # type: ignore
+
+        # Nested looping + conditionals are necessary for the drawing of each light
+        # source. Ignored relevant style checks for that reason.
         for b in range(batch_size):
             n_sources = int(full_cat.n_sources[b].item())
-            gs_img = galsim.Image(array=images[b, 0], scale=self.pixel_scale)
-            for s in range(n_sources):
-                if full_cat["galaxy_bools"][b][s] == 1:
-                    galsim_obj = self.render_galaxy(full_cat["galaxy_params"][b][s])
-                elif full_cat["star_bools"][b][s] == 1:
-                    galsim_obj = self.psf_galsim.withFlux(full_cat["star_fluxes"][b][s].item())
-                else:
-                    raise AssertionError("Every source is a star or galaxy")
-
-                plocs = full_cat.plocs[b][s]
-                offset = np.array([plocs[1] - (slen_w / 2), plocs[0] - (slen_h / 2)])
-                # essentially all the runtime of the simulator is incurred by this call to drawImage
-                galsim_obj.drawImage(offset=offset, method="auto", add_to_image=True, image=gs_img)
+            psf = psfs[b]
+            for bnd in range(self.n_bands):
+                gs_img = galsim.Image(array=images[b, bnd], scale=self.pixel_scale)
+                for s in range(n_sources):
+                    if full_cat["galaxy_bools"][b][s] == 1:
+                        gal_with_band = torch.cat(  # noqa: WPS220,WPS317
+                            (
+                                torch.unsqueeze(full_cat["galaxy_params"][b][s][bnd], 0),
+                                full_cat["galaxy_params"][b][s][self.n_bands :],
+                            ),
+                            0,
+                        )
```

Comment:
Would you write the lines above a bit more concisely, i.e. using some temp variables? The way black autoformats some things is quite hard to read.